File: src/Binary/SearchRotatedSortedArray.java
package Binary;

// https://leetcode.com/problems/search-in-rotated-sorted-array/
public class SearchRotatedSortedArray {
    public int search(int[] nums, int target) {
        int smallestIndex = findSmallestIndex(nums);

        if (nums[smallestIndex] == target) {
            return smallestIndex;
        }
        int low = 0;
        int high = nums.length - 1;
        if (target <= nums[nums.length - 1]) {
            low = smallestIndex + 1;
        } else {
            high = smallestIndex - 1;
        }

        while (low <= high) { // standard binary search
            int mid = (low + high) / 2;
            if (nums[mid] == target) return mid;
            else if (target > nums[mid]) low = mid + 1;
            else high = mid - 1;
        }
        return -1;
    }

    public int findSmallestIndex(int[] nums) {
        int low = 0;
        int high = nums.length - 1;
        while (high > low) {
             int mid = (high + low) / 2;
             if (nums[mid] > nums[high]) {
                 low = mid + 1;
             } else {
                 high = mid;
             }
        }
        return low;
    }

    public static void main(String[] args) {
        SearchRotatedSortedArray s = new SearchRotatedSortedArray();
        int nums[] = {4,5,6,7,8,0,1,2,3};
        System.out.println("index is: " + s.search(nums, 3));
    }
}

File: py/essential/SearchRotatedSortedArray.py
# Blind
# https://leetcode.com/problems/search-in-rotated-sorted-array/description/?envType=problem-list-v2&envId=oizxjoit


from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        res, n = -1, len(nums)

        def findSmallestIndex():
            left, right = 0, n - 1

            while left < right:
                mid = (left + right) // 2
                if nums[mid] < nums[right]:
                    right = mid
                else:
                    left = mid + 1
            return left

        smallestIndex = findSmallestIndex()
        if nums[smallestIndex] == target:
            return smallestIndex
        left = 0
        right = n - 1
        if nums[right] > target:
            left = smallestIndex + 1
        elif nums[right] < target:
            right = smallestIndex - 1
        else:
            return right

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return res


s = Solution()
print(s.search([4, 5, 6, 7, 0, 1, 2], 7))
